Apache kafka (#1712)
knative/docs: docs/eventing/samples/kafka/source/README.md

---
title: "Apache Kafka Source Example"
linkTitle: "Source - Knative service"
weight: 10
type: "docs"
---

Tutorial on how to build and deploy a `KafkaSource` [Eventing source](../../../sources/README.md) using a Knative Serving `Service`.

## Prerequisites

You must ensure that you meet the [prerequisites listed in the Apache Kafka overview](../README.md).

## Creating a `KafkaSource` source CRD

1. Install the `KafkaSource` sub-component to your Knative cluster:
   ```

>    kubectl apply -f https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-importer.yaml

Reviewer: Should this use 0.9.0?

     topics: knative-demo-topic
     sink:
       apiVersion: serving.knative.dev/v1alpha1
       kind: Service
       name: event-display
   ```

1. Deploy the event source.
   ```
   $ kubectl apply -f kafka/source/samples/event-source.yaml
   ...
   kafkasource.sources.eventing.knative.dev/kafka-source created
   ```
1. Check that the event source pod is running. The pod name will be prefixed
   with `kafka-source`.
   ```
   $ kubectl get pods
   NAME                                  READY     STATUS    RESTARTS   AGE
   kafka-source-xlnhq-5544766765-dnl5s   1/1       Running   0          40m
   ```
1. Ensure the Apache Kafka Event Source started with the necessary
   configuration.
   ```
   $ kubectl logs --selector='knative-eventing-source-name=kafka-source'
   {"level":"info","ts":"2019-04-01T19:09:32.164Z","caller":"receive_adapter/main.go:97","msg":"Starting Apache Kafka Receive Adapter...","Bootstrap Server":"...","Topics":".","ConsumerGroup":"...","SinkURI":"...","TLS":false}
   ```

### Verify

1. Produce a message (`{"msg": "This is a test!"}`) to the Apache Kafka topic, like shown below:
   ```

>    kubectl -n kafka run kafka-produc1er -ti --image=strimzi/kafka:0.13.0-kafka-2.3.0 --rm=true --restart=Never -- bin/kafka-console-producer.sh --broker-list my-cluster-kafka-bootstrap:9092 --topic my-topic

Reviewer: Is this supposed to be `kafka-produc1er` or is that a typo?

   If you don't see a command prompt, try pressing enter.
   >{"msg": "This is a test!"}
   ```
1. Check that the Apache Kafka Event Source consumed the message and sent it to
   its sink properly.
   ```
   $ kubectl logs --selector='knative-eventing-source-name=kafka-source'
   ...
   {"level":"info","ts":"2019-04-15T20:37:24.702Z","caller":"receive_adapter/main.go:99","msg":"Starting Apache Kafka Receive Adapter...","bootstrap_server":"...","Topics":"knative-demo-topic","ConsumerGroup":"knative-group","SinkURI":"...","TLS":false}
   {"level":"info","ts":"2019-04-15T20:37:24.702Z","caller":"adapter/adapter.go:100","msg":"Starting with config: ","bootstrap_server":"...","Topics":"knative-demo-topic","ConsumerGroup":"knative-group","SinkURI":"...","TLS":false}
   {"level":"info","ts":1553034726.546107,"caller":"adapter/adapter.go:154","msg":"Successfully sent event to sink"}
   ```

1. Ensure the Event Display received the message sent to it by the Event Source.
   ```

>    $ kubectl logs event-display-00001-deployment-5d5df6c7-gv2j4 -c user-container

Reviewer: This should use a label selector instead of a pod name, like `--selector='serving.knative.dev/service: event-display'`

   2019/03/19 22:25:54 Starting Apache Kafka controller.
   ```

#### Apache Kafka Topic (Optional)

1. If using Strimzi, you can can set a topic modifying
   `source/kafka-topic.yaml` with your desired:

- Topic
- Cluster Name
- Partitions
- Replicas

  ```yaml
  apiVersion: kafka.strimzi.io/v1alpha1
  kind: KafkaTopic
  metadata:
    name: knative-demo-topic
    namespace: kafka
    labels:
      strimzi.io/cluster: my-cluster
  spec:
    partitions: 3
    replicas: 1
    config:
      retention.ms: 7200000
      segment.bytes: 1073741824
  ```

2. Deploy the `KafkaTopic`

   ```shell
   $ kubectl apply -f kafka/source/samples/strimzi-topic.yaml
   kafkatopic.kafka.strimzi.io/knative-demo-topic created
   ```

3. Ensure the `KafkaTopic` is running.

   ```shell
   $ kubectl -n kafka get kafkatopics.kafka.strimzi.io
   NAME                 AGE
   knative-demo-topic   16s
   ```

#### Event Display

1. Build and deploy the Event Display Service.
   ```
   $ kubectl apply -f source/event-display.yaml
   ...
   service.serving.knative.dev/event-display created
   ```
2. Ensure that the Service pod is running. The pod name will be prefixed with
   `event-display`.
   ```

>    $ kubectl get pods

Reviewer 2: I failed here with `No resources found`

Note: I skipped the optional Strimzi steps. If i run the following it says the service is not running:

```
kubectl get service.serving.knative.dev
NAME            URL                                        LATESTCREATED         LATESTREADY   READY   REASON
event-display   http://event-display.default.example.com   event-display-f8c8d                 False   RevisionMissing
```

Author: hrm, weird

Reviewer 3: That may be problem with istio/serving setup?

# Apache Kafka - Source Example

> Tutorial on how to build and deploy a `KafkaSource` [Eventing source](../../../sources/README.md) using a Knative Serving `Service`.

Reviewer 2: ```suggestion
Learn how to build and deploy a `KafkaSource` [Eventing source](../../../sources/README.md) using a Knative Serving `Service`.
```

## Prerequisites

You must ensure that you meet the [prerequisites listed in the Apache Kafka overview](../README.md).

## Creating a `KafkaSource` source CRD

1. Install the `KafkaSource` sub-component to your Knative cluster:
   ```
   kubectl apply -f https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-importer.yaml
   ```
2. Check that the `kafka-controller-manager-0` pod is running.
   ```
   kubectl get pods --namespace knative-sources
   NAME                         READY     STATUS    RESTARTS   AGE
   kafka-controller-manager-0   1/1       Running   0          42m
   ```
3. Check the `kafka-controller-manager-0` pod logs.
   ```
   $ kubectl logs kafka-controller-manager-0 -n knative-sources
   2019/03/19 22:25:54 Registering Components.
   2019/03/19 22:25:54 Setting up Controller.
   2019/03/19 22:25:54 Adding the Apache Kafka Source controller.
   2019/03/19 22:25:54 Starting Apache Kafka controller.
   ```

> ### Apache Kafka Topic (Optional)

Reviewer 2: ```suggestion
### (Strimzi only) Set an Apache Kafka Topic

Run the following commands if you are running an Apache Kafka cluster using Strimzi:
```

  spec:
    partitions: 3
    replicas: 1
    config:
      retention.ms: 7200000
      segment.bytes: 1073741824
  ```

2. Deploy the `KafkaTopic`

   ```shell
   $ kubectl apply -f kafka/source/samples/strimzi-topic.yaml
   kafkatopic.kafka.strimzi.io/knative-demo-topic created
   ```

3. Ensure the `KafkaTopic` is running.

   ```shell
   $ kubectl -n kafka get kafkatopics.kafka.strimzi.io
   NAME                 AGE
   knative-demo-topic   16s
   ```

### Create the Event Display service

1. Build and deploy the Event Display Service.
   ```
   $ kubectl apply --filename source/samples/event-display.yaml
   ...
   service.serving.knative.dev/event-display created
   ```
2. Ensure that the Service pod is running. The pod name will be prefixed with
   `event-display`.
   ```
   $ kubectl get pods
   NAME                                            READY     STATUS    RESTARTS   AGE
   event-display-00001-deployment-5d5df6c7-gv2j4   2/2       Running   0          72s
   ...
   ```

#### Apache Kafka Event Source

1. Modify `source/event-source.yaml` accordingly with bootstrap
   servers, topics, etc...

   NOTE: If using an internal Apache Kafka cluster, you may need to ensure

>    you've specified the correct variables set in any `KafkaTopic` and

Reviewer 4: What's KafkaTopic here? I don't see it in the example, nor event-source

   `event-source`. For example, the following source could be used:

   ```yaml
   apiVersion: sources.eventing.knative.dev/v1alpha1
   kind: KafkaSource
   metadata:
     name: kafka-source
   spec:
     consumerGroup: knative-group
     bootstrapServers: my-cluster-kafka-bootstrap.kafka:9092 #note the kafka namespace
     topics: knative-demo-topic
     sink:
       apiVersion: serving.knative.dev/v1alpha1
       kind: Service
       name: event-display
   ```

> 2. Build and deploy the event source.

Reviewer 4: I think this is just 1.

> 2. Build and deploy the event source.

Reviewer 4: This doesn't build anything? Just deploys?

   ```
   $ kubectl apply -f kafka/source/samples/event-source.yaml
   ...
   kafkasource.sources.eventing.knative.dev/kafka-source created
   ```

> 3. Check that the event source pod is running. The pod name will be prefixed

Reviewer 4: I think this is just 1.

   with `kafka-source`.
   ```
   $ kubectl get pods
   NAME                                  READY     STATUS    RESTARTS   AGE
   kafka-source-xlnhq-5544766765-dnl5s   1/1       Running   0          40m
   ```

> 4. Ensure the Apache Kafka Event Source started with the necessary

Reviewer 4: just 1.
Is there a label that we could use for logs? As in kubectl logs -selector...?

   configuration.
   ```
   $ kubectl logs kafka-source-xlnhq-5544766765-dnl5s
   {"level":"info","ts":"2019-04-01T19:09:32.164Z","caller":"receive_adapter/main.go:97","msg":"Starting Apache Kafka Receive Adapter...","Bootstrap Server":"...","Topics":".","ConsumerGroup":"...","SinkURI":"...","TLS":false}
   ```

### Verify

> 1. Produce the message shown below to Apache Kafka.

Reviewer 4: How do you do this?

   ```
   {"msg": "This is a test!"}
   ```
2. Check that the Apache Kafka Event Source consumed the message and sent it to
   its sink properly.

   ```
   $ kubectl logs kafka-source-xlnhq-5544766765-dnl5s
   ...
   {"level":"info","ts":"2019-04-15T20:37:24.702Z","caller":"receive_adapter/main.go:99","msg":"Starting Apache Kafka Receive Adapter...","bootstrap_server":"...","Topics":"knative-demo-topic","ConsumerGroup":"knative-group","SinkURI":"...","TLS":false}
   {"level":"info","ts":"2019-04-15T20:37:24.702Z","caller":"adapter/adapter.go:100","msg":"Starting with config: ","bootstrap_server":"...","Topics":"knative-demo-topic","ConsumerGroup":"knative-group","SinkURI":"...","TLS":false}
   {"level":"info","ts":1553034726.546107,"caller":"adapter/adapter.go:154","msg":"Successfully sent event to sink"}
   ```

> 3. Ensure the Event Display received the message sent to it by the Event Source.

Reviewer 4: same here, can we use label selector so it's just a cut&paste for the user?

   ```

>    $ kubectl logs --selector='serving.knative.dev/service: event-display' -c user-container

Reviewer: Oops now that I'm looking at this again, I see it was the wrong syntax. Label selectors use `=` as a separator. Sorry about that :disappointed:

`--selector='serving.knative.dev/service=event-display'`

     Time: 2019-03-19T22:32:06.535321588Z
     ContentType: application/json
     Extensions:
       key:
   Transport Context,
     URI: /
     Host: event-display.default.svc.cluster.local
     Method: POST
   Data,
     {
       "msg": "This is a test!"
     }
   ```

## Teardown Steps

1. Remove the Apache Kafka Event Source
   ```
   $ kubectl delete -f source/source.yaml
   kafkasource.sources.eventing.knative.dev "kafka-source" deleted
   ```
2. Remove the Event Display
   ```
   $ kubectl delete -f source/event-display.yaml
   service.serving.knative.dev "event-display" deleted
   ```
3. Remove the Apache Kafka Event Controller
   ```
   $ kubectl delete -f https://github.com/knative/eventing-contrib/releases/download/v0.8.2/kafka-importer.yaml
   serviceaccount "kafka-controller-manager" deleted
   clusterrole.rbac.authorization.k8s.io "eventing-sources-kafka-controller" deleted
   clusterrolebinding.rbac.authorization.k8s.io "eventing-sources-kafka-controller" deleted
   customresourcedefinition.apiextensions.k8s.io "kafkasources.sources.eventing.knative.dev" deleted
   service "kafka-controller" deleted
   statefulset.apps "kafka-controller-manager" deleted
   ```

> 4. (Optional) Remove the Apache Kafka Topic

Reviewer 2: ```suggestion
4. (Strimzi only) Remove the Apache Kafka Topic
```

> 

Reviewer 2: ```suggestion

Run the following command if you are running an Apache Kafka cluster using Strimzi:

```
